Mathlib/Data/Multiset/DershowitzManna.lean

```diff
@@ -0,0 +1,410 @@
+/-
+Copyright (c) 2024 Haitian Wang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Haitian Wang, Malvin Gattinger
+-/
+import Mathlib.Tactic.Linarith
+import Mathlib.Data.Multiset.Basic
+import Mathlib.Logic.Relation
+
+/-!
+# Dershowitz-Manna ordering
+
+In this file we define the _Dershowitz-Manna ordering_ on multisets. Specifically, for two multisets
+`M` and `N` in a partial order `(S, <)`, `M` is smaller than `N` in the Dershowitz-Manna ordering if
+`M` can be obtained from `N` by replacing one or more elements in `N` by some finite number of
+elements from `S`, each of which is smaller (in the underling ordering over `S`) than one of the
+replaced elements from `N`. We prove that, given a well-founded partial order on the underlying set,
+the Dershowitz-Manna ordering defined over multisets is also well-founded.
+
+## Main results
+
+- `Multiset.IsDershowitzMannaLT` : the standard definition fo the `Dershowitz-Manna ordering`.
+- `Multiset.wellFounded_isDershowitzMannaLT` : the main theorem about the
+`Dershowitz-Manna ordering` being well-founded.
+- `Multiset.TransLT_eq_DMLT` : two definitions of the `Dershowitz-Manna ordering` are equivalent.
+
+## References
+
+* [Wikipedia, Dershowitz–Manna ordering*]
+(https://en.wikipedia.org/wiki/Dershowitz%E2%80%93Manna_ordering)
+
+* [CoLoR](https://github.com/fblanqui/color), a Coq library on rewriting theory and termination.
+  Our code here is inspired by their formalization and the theorem is called `mOrd_wf` in the file
+  [MultisetList.v](https://github.com/fblanqui/color/blob/1.8.5/Util/Multiset/MultisetOrder.v).
+
+-/
+
+namespace Multiset
+
+variable {α : Type*} [Preorder α]
+
+/-- The standard Dershowitz–Manna ordering. -/
+def IsDershowitzMannaLT (M N : Multiset α) : Prop :=
+  ∃ (X Y Z : Multiset α),
+      Z ≠ ∅
+    ∧ M = X + Y
+    ∧ N = X + Z
+    ∧ (∀ y ∈ Y, ∃ z ∈ Z, y < z)
+
+/-- A special case of `IsDershowitzMannaLT`. The transitive closure of it is used to define
+an equivalent (proved later) version of the ordering. -/
+private def OneStepLT [LT α] (M N : Multiset α) : Prop :=
+  ∃ X Y a,
+      M = X + Y
+    ∧ N = X + {a}
+    ∧ ∀ y ∈ Y, y < a
+
+open Relation
+
+/-- The transitive closure of `OneStepLT` and is equivalent to `IsDershowitzMannaLT`
+(proved later). -/
+private def TransLT [LT α] : Multiset α → Multiset α → Prop := TransGen OneStepLT
+
+/-- A special case of `IsDershowitzMannaLT`. -/
+private lemma dmlt_of_DMLT_singleton (M N : Multiset α)
```

Comment:
Again, not crucial, because they are private, but can you rename your auxiliary lemmas to somewhat match the naming convention? Eg
```suggestion
private lemma isDershowitzMannaLT_of_oneStepLT {M N : Multiset α}
```